Custom profile (#20)
alliander-opensource/pycgmes: README.md

SPDX-License-Identifier: Apache-2.0
-->

# PyCGMES

> <<<<<<< HEAD

Reviewer: ```suggestion
```

        - [Attribute namespace](#attribute-namespace)
  - [Content of this repository](#content-of-this-repository)
    - [Schemas v3](#schemas-v3)
    - [SHACL files](#shacl-files)
    - [V3 source zip](#v3-source-zip)
    - [Dataclasses](#dataclasses)
  - [Library build, CI, CD...](#library-build-ci-cd)
    - [CI](#ci)
    - [CD](#cd)
  - [License](#license)

> >>>>>>> 4384d02 (update docs with CGMES info)

Reviewer: ```suggestion
```

Author: 🤦‍♂️

## Custom attributes

You might want to add extra attributes. For instance, the color of a cable (ACLineSegment). This is possible in 2 ways.

You can look at the [examples](./examples/custom_attributes.py)

### Add to an existing profile

If you need to add your own attributes (example: cable colour), you can do that by subclassing the relevant class, and
add a (or many) new atributes there.

If this is a leaf node (for instance `ACLineSegment`), it "just works". If you want to add an extra attribute to a
class higher in the hierarchy (for instance `Equipment`) there is a lot more work to do.

```python
@dataclass(config=DataclassConfig)
class CustomBay(Bay):
    colour: str = Field(
        default="Red",
        in_profiles=[
            Profile.EQ,
        ],
    )
```

### Create a new profile

This is cleaner - the official CGMES profiles stays untouched, and a new profile is delivered next to it.

You can do this by extending the BaseProfile Enum in [Profile.py](./pycgmes/utils/profile.py).

In Python, it is not possible to extend or compose Enums which already have fields, but you can just create your own:

```python
from pycgmes.utils.profile import BaseProfile

> class CustomProfile(BaseProfile):

Author: You are totally correct! I will look into that.

    CUS="Tom"
    FRO="Mage"
```

And use it everywhere you would use a profile:

```python
from pycgmes.utils.dataclassconfig import DataclassConfig

@dataclass(config=DataclassConfig)
class CustomBayAttr(Bay):
    colour: str = Field(
        default="Red",
        in_profiles=[
            CustomProfile.CUS,
        ],
    )

# And for instance:
custom_attrs = CustomBayAttr(colour="purple").cgmes_attributes_in_profile(CustomProfile.CUS)
```

### Implementation details

#### Apparent class

> By default, an attribute is fully qualified. `bch` in `ACLineSegment` will appear as `ACLineSegment.bch` in the serialisation.

Reviewer: ```suggestion
By default, an attribute is fully qualified. `attribute` in `ACLineSegment` will appear as `ACLineSegment.attribute` in the serialisation.
```

> For a custom attribute, you might not want to see  `ACLineSegmentCustom.bch`. To prevent this, you can override the `apparent_name`
> of your custom class:

Reviewer: ```suggestion
In the case of a custom attribute defined via a sub class, the result would be: `ACLineSegmentCustom.customAttribute`. To preserve the original class name (i.e. serialise your attribute as `ACLineSegment.customAttribute`), you need to override the `apparent_name` of your custom class:
```

        return "ACLineSegment"
```

#### Namespace

In the serialisation, the namespace of all attributes is `cim` (`"http://iec.ch/TC57/2013/CIM-schema-cim16#"`) by default.

> The serialisation is not done by PyCGMES (yet), but if you want a custom namespace for an attribute,

Reviewer: This works for namespaces of classes, but 1) it implies that if you define a new profile, you have to specify namespace for every attribute you attach from that namespace; 2) I am not sure how it would work to define the namespace for a custom class that belongs to a custom profile.

Author: 1) this is currently correct. 
2) classes do not have namespaces, only attributes if I understood correctly?

If I understand it correctly, you would like to have a way to set a "class namespace", which would then apply to all its attribute? That feels like a good idea.

Reviewer: following what's defined in: https://www.entsoe.eu/Documents/CIM_documents/Grid_Model_CIM/CGMES_Profiling_User_Guide_v1.0.pdf

when you create an extension of CIM (profiles, classes, attributes), you usually define a namespace for it.

this is actually very convenient naming wise, because you could define a custom ACLineSegment as `my:ACLineSegment` instead of `cim:CustomACLineSegment`

Author: I see. I had missed this indeed. It should be an easy fix.
